promql/query_logger.go

```diff
@@ -108,7 +128,7 @@ func getMMapedFile(filename string, filesize int, logger log.Logger) ([]byte, io
 		return nil, nil, err
 	}
 
-	return fileAsBytes, file, err
+	return fileAsBytes, &mmapedFile{f: file, m: fileAsBytes}, err
```

Comment:
I think Ben's suggestion here https://github.com/prometheus/prometheus/pull/13948#discussion_r1577884572
was that we can even close the file descriptor in `getMMapedFile` (thus we no longer need to return it)
Did you try that? Doesn't it work on windows?

Reply:
I don't have Windows, so I'm not able to test there myself. I think Ben only found documentation for Linux though, not Unix in general. Are you able to find corresponding documentation for other Unices?

Reply:
Reading [documentation](https://learn.microsoft.com/en-us/windows/win32/memory/creating-a-file-mapping-object) on mmapping on Windows, it looks as if one should keep the file open for the lifetime of the mapping:

> In addition, the file handle should remain open until the process no longer needs the file mapping object.

It also says the following about closing the mmapped file:

> When each process finishes using the file mapping object and has unmapped all views, it must close the file mapping object's handle and the file on disk by calling CloseHandle.

Reply:
I was implicitly suggesting to give it a try on the windows job in the CI (You could still give it a try to confirm the statements in the doc if you want)

> Based on the above, I believe we should keep the current behaviour (of the PR). I also don't see that we gain much by closing the file early?

Just for less/simpler code, as we are now getting closer and closer to https://github.com/prometheus/prometheus/blob/1081e336a0c4004b6f817c70f0ff4f3810db80fa/tsdb/fileutil/mmap.go#L21, we’ll see if we can merge them in the future.

Reply:
My take is that even if it would work in CI, we shouldn't go against what the Windows mmap documentation specifies.

Reply:
I did a test with CI, and closing the file before the mmap handle actually does fail: `mmapedFile: unmapping: FlushFileBuffers: The handle is invalid.`.